src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/RenameTypeService.cs

```diff
@@ -0,0 +1,239 @@
+                {
+                    project = proj;
+                    return true;
+                }
+
+            }
+
+            project = null;
+            return false;
+        }
+
+        private async Task<bool> CanHandleRenameAsync(string oldName, string newName, Project project)
+        {
+            (bool success, bool isCaseSensitive) = await TryDetermineIfCompilationIsCaseSensitiveAsync(project);
+            if (!success)
+                return false;
+
+            return IsValidIdentifier(oldName) &&
+                   IsValidIdentifier(newName) &&
+                       (!string.Equals(
+                            oldName,
+                            newName,
+                            isCaseSensitive
+                                ? StringComparison.Ordinal
+                                : StringComparison.OrdinalIgnoreCase));
+        }
+
+        private bool IsValidIdentifier(string identifierName)
+        {
+            return SyntaxFactsService?.IsValidIdentifier(identifierName) ?? false;
+        }
+
+        private static async Task<(bool success, bool isCaseSensitive)> TryDetermineIfCompilationIsCaseSensitiveAsync(Project project)
+        {
+            Compilation compilation = await project.GetCompilationAsync();
+            if (compilation is null)
+            {
+                // this project does not support compilations
+                return (false, false);
+            }
+
+            return (true, compilation.IsCaseSensitive);
+        }
+
+        private static bool  TryGetCodeElementToRename(string oldName,
+                                                       EnvDTE.CodeElements elements,
+                                                       bool isCaseSensitive,
+                                                       out EnvDTE80.CodeElement2 codeElementToRename)
+        {
+            foreach (EnvDTE.CodeElement element in elements)
+            {
+                if (element.Kind == EnvDTE.vsCMElement.vsCMElementNamespace)
+                {
+                    if (TryGetCodeElementToRename(oldName, element.Children, isCaseSensitive, out codeElementToRename))
+                    {
+                        return true;
+                    }
+                }
+
+                if (element.Kind == EnvDTE.vsCMElement.vsCMElementStruct ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementClass ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementModule ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementInterface ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementEnum)
+                {
+                    string elementName = element.Name;
+                    if(string.Equals(oldName, elementName, isCaseSensitive ? StringComparison.Ordinal: StringComparison.OrdinalIgnoreCase))
+                    {
+                        codeElementToRename = element as EnvDTE80.CodeElement2;
```

Comment:
`as` seems dangerous: if this could somehow not be a CodeElement2 you're going to nullref at the caller.